bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BValueArray.java

```diff
@@ -0,0 +1,573 @@
+            }
+        }
+        return sj.toString();
+    }
+
+    @Override
+    public BValue getBValue(long index) {
+        if (elementType != null) {
+            if (elementType.getTag() == TypeTags.INT_TAG) {
+                return new BInteger(getInt(index));
+            } else if (elementType.getTag() == TypeTags.BOOLEAN_TAG) {
+                return new BBoolean(getBoolean(index) == 1);
+            } else if (elementType.getTag() == TypeTags.BYTE_TAG) {
+                return new BByte(getByte(index));
+            } else if (elementType.getTag() == TypeTags.FLOAT_TAG) {
+                return new BFloat(getFloat(index));
+            } else if (elementType.getTag() == TypeTags.STRING_TAG) {
+                return new BString(getString(index));
+            } else {
+                return getRefValue(index);
+            }
+        }
+        return getRefValue(index);
+    }
+
+    public BRefType<?>[] getValues() {
+        return refValues;
+    }
+
+    public byte[] getBytes() {
+        return byteValues.clone();
+    }
+
+    @SuppressWarnings("unchecked")
+    public String[] getStringArray() {
+        return stringValues;
+    }
+
+    @Override
+    public String toString() {
+        return stringValue();
+    }
+
+    private String getJSONString() {
+        ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
+        JsonGenerator gen = new JsonGenerator(byteOut);
+        try {
+            gen.serialize(this);
+            gen.flush();
+        } catch (IOException e) {
+            throw new BallerinaException("Error in converting JSON to a string: " + e.getMessage(), e);
+        }
+        return new String(byteOut.toByteArray());
+    }
+
+    private BType getElementType(BType type) {
+        if (type.getTag() != TypeTags.ARRAY_TAG) {
+            return type;
+        }
+
+        return getElementType(((BArrayType) type).getElementType());
+    }
+
+    @Override
+    public SerializableRefType serialize(SerializableState state) {
+
+        return new SerializableBRefArray(this, state);
+    }
+
+    @Override
+    public void serialize(OutputStream outputStream) {
+        if (elementType.getTag() == TypeTags.BYTE_TAG) {
+            try {
+                outputStream.write(byteValues);
+            } catch (IOException e) {
+                throw new BallerinaException("error occurred while writing the binary content to the output stream", e);
+            }
+        } else {
+            try {
+                outputStream.write(this.stringValue().getBytes(Charset.defaultCharset()));
+            } catch (IOException e) {
+                throw new BallerinaException("error occurred while serializing data", e);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void attemptFreeze(CPU.FreezeStatus freezeStatus) {
+        if (isOpenForFreeze(this.freezeStatus, freezeStatus)) {
```

Comment:
shall we invert the if condition, and fail fast?

Reply:
Yes, done.